controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,177 @@
+    std::unique_lock<std::mutex> lock(async_mtx_, std::try_to_lock);
+    if (!async_update_ready_ && lock.owns_lock())
+    {
+      async_update_ready_ = true;
+      current_update_time_ = time;
+      current_update_period_ = period;
+      lock.unlock();
+      async_update_condition_.notify_one();
+    }
+    return async_update_return_;
+  }
+
+  void wait_for_update_to_finish()
+  {
+    if (is_running())
+    {
+      std::unique_lock<std::mutex> lock(async_mtx_);
+      async_update_condition_.wait(lock, [this] { return !async_update_ready_; });
+    }
+  }
+
+  bool is_initialized() const
+  {
+    return get_state_function_ != nullptr && async_function_ != nullptr;
+  }
+
+  void join_async_update_thread()
+  {
+    if (is_running())
+    {
+      thread_.join();
+    }
+  }
+
+  bool is_running() const { return thread_.joinable(); }
+
+  bool is_preempted() const
+  {
+    return (
+      async_update_stop_ ||
+      (is_initialized() &&
+       get_state_function_().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE));
+  }
+
+  void preempt_async_update()
+  {
+    if (is_running())
+    {
+      async_update_stop_ = true;
+      async_update_condition_.notify_one();
+      thread_.join();
+    }
+  }
+
+private:
+  void initialize_async_update_thread(
+    std::function<const rclcpp_lifecycle::State &()> get_state_function,
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function)
+  {
+    // * If the thread is not joinable, create a new thread and call the update function and wait
+    // for it to be triggered again and repeat the same process.
+    // * If the thread is joinable, check if the current update cycle is finished. If so, then
+    // trigger a new update cycle.
+    // * The controller managr is responsible for triggering the update cycle and maintaining the
+    // controller's update rate and should be acting as a scheduler.
+    if (get_state_function_ == nullptr || async_function_ == nullptr)
+    {
+      throw std::runtime_error("AsyncFunctionHandler not initialized properly!");
+    }
+    if (!thread_.joinable())
+    {
+      async_update_stop_ = false;
+      async_update_ready_ = false;
+      async_update_return_ = T();
+      thread_ = std::thread(
+        [this]() -> void
+        {
+          // \note There might be an concurrency issue with the get|ate() call here. This mightn't
+          // be critical here as the state of the controller is not expected to change during the
+          // update cycle
+          while (get_state_function_().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE &&
+                 !async_update_stop_)
+          {
+            std::unique_lock<std::mutex> lock(async_mtx_);
+            async_update_condition_.wait(
+              lock, [this] { return async_update_ready_ || async_update_stop_; });
+            if (async_update_stop_)
+            {
+              lock.unlock();
```

Comment:
Don't need this. The break below would cause the lock to go out of scope.

Reply:
Sure.

Reply:
Done